File: personTests/pet.js
function Pet(name, age) {
    this.name = name || "";
    this.age = age | 0;
    
    this.isAlive = function () {
        return this.age < 20;
    };
    
    this.isBaby = function () {
        return this.age < 2;
    };
    
    this.haveBirthday = function () {
        this.age += 1;
    }
}

File: VisualStudio/UnitTestDemo/Entities/Pet.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Entities
{
    public class Pet
    {
        public Pet(string name, int age)
        {
            Name = name;
            Age = age;
        }

        private string name;
        public string Name
        {
            get { return name; }
            private set
            {
                if (string.IsNullOrEmpty(value))
                {
                    throw new ArgumentException("Name is null or empty!");
                }

                name = value;
            }
        }

        private int age;
        public int Age
        {
            get { return age; }
            private set
            {
                if (value < 0 || value > 20)
                {
                    throw new PetAgeException();
                }
                
                age = value;
            }
        }

        public IPetDatabase Database { get; set; }

        public bool SaveToDatabase()
        {
            return Database.SaveNewPet(this);
        }


    }

    public class PetAgeException : ArgumentOutOfRangeException
    {
        public PetAgeException() : base("Age is not in the valid range [0-20]")
        {

        }
    }
}
